src/cast_core/src/CrudActor.cpp

```diff
@@ -1008,7 +1008,9 @@ struct WithTransactionOperation : public BaseOperation {
         }
         auto createOp = opConstructor->second;
         auto& yamlCommand = txnOp["OperationCommand"];
-        _txnOps.push_back(createOp(yamlCommand, _onSession, _collection, _operation, context, id));
+        // operations can override withTransaction level
+        bool onSession = yamlCommand["OnSession"].maybe<bool>().value_or(_onSession);
+        _txnOps.push_back(createOp(yamlCommand, onSession, _collection, _operation, context, id));
```

Comment:
Since this isn't tested (admittedly hard to do) let's add a comment here for future maintainers. Something like

```
// This behavior is hard to test.
// The CrudActorYamlTests suite exercises both branches of this boolean but doesn't assert the behavior.
// Be careful when making changes around this code.
```

Also I wonder if it would be possible to test by hand. Write a dummy workload that does a whole bunch of inserts on a session and compare the behavior with a similar workload without the session? It's almost certainly not going to be worth the effort for this one bug, but that may be an interesting avenue to explore in the future.

Reply:
And yes, we intentionally did not invest in lots of machinery to unit-test Actors. Feedback or ideas welcome, but it quickly went into template/macro meta-programming hell to allow for monkey-patching at test time. It was beyond my expertise so just decided to have an integration testing setup that was as robust and easy-to-use as possible. Transactions and failure cases are admittedly a blind-spot in this approach.